include/systems/generic_projector.h

```diff
@@ -672,10 +662,8 @@ void GenericProjector<FFunctor, GFunctor, FValue, ProjectionAction>::operator()
 
       // Loop over all the variables we've been requested to project, to
       // do the projection
-      for (std::size_t v=0; v!=variables.size(); v++)
+      for (const auto & var : variables)
```

Comment:
I know it's trivial outside of the innermost inner loops, but - is there a performance implication to using a reference rather than a copy here?

The reference will be a pointer under the hood, so unless the compiler can prove that the pointee never changes it will have to hang on to and dereference 8 bytes rather than just hanging on to 4.

And we have a bunch of function calls, including one that I'm sure will never be inlined, so unless the compiler does ridiculous levels of inter-procedural optimization it won't be able to prove that those calls don't change the variables container in mid-iteration.

Reply:
I know, this particular example called size() on every loop iteration, so the reference-based range for is probably *still* faster, but in theory it seems like removing the & would buy a couple more nanoseconds on top of that.

Reply:
I always prefer `const auto &` because somebody could potentially come along later and change the underlying type in `variables` to e.g. a `shared_ptr<T>`, in which case `const auto foo` would then trigger a reference-counted copy construction (and destruction) at each iteration of the loop. Even worse would be if the underlying type changed to simply `T`...

I'm willing to be convinced that there's a performance difference between `const auto` and `const auto &` for the case of dumb pointers, but I don't think it's worth the possibility of introducing a different bug or different performance penalty due to a later refactoring down the line...

Reply:
> Even worse would be if the underlying type changed to simply T...

By this I meant some case where `T` is expensive to copy, not the `unsigned int` case here!